lib/improver/weights.py

```diff
@@ -30,123 +30,259 @@
+                y0val : Positive float
+                        relative value of starting point. Default = 1.0
 
-    return weights
+                AND EITHER:
+                slope : float
+                        slope of the line. Default = 0.0 (equal weights)
+                OR
+                ynval : Positive float or None
+                        Relative weights of last point.
+                        Default value is None
+
+            Returns:
+                weights : array of weights
+                          sum of all weights = 1.0
+
+        """
+        if not isinstance(num_of_weights, int) or num_of_weights <= 0:
+            msg = ('Number of weights must be integer > 0 '
+                   'num = {0:s}'.format(str(num_of_weights)))
+            raise ValueError(msg)
+        # Special case num_of_weighs == 1 i.e. Scalar coordinate.
+        if num_of_weights == 1:
+            weights = np.array([1.0])
+            return weights
+        if not isinstance(y0val, float) or y0val <= 0.0:
+            msg = ('y0val must be a float > 0.0, '
+                   'y0val = {0:s}'.format(str(y0val)))
+            raise ValueError(msg)
+        if ynval is not None:
+            if slope != 0.0:
+                msg = ('Relative end point weight or slope must be set'
+                       ' but not both.')
+                raise ValueError(msg)
+            else:
+                slope = (ynval - y0val)/(num_of_weights - 1.0)
+
+        weights_list = []
+        for tval in range(0, num_of_weights):
+            weights_list.append(slope*tval + y0val)
+
+        weights = WeightsUtilities.normalise_weights(np.array(weights_list))
+
+        return weights
+
+    @staticmethod
+    def process_coord(cube, coord):
+        """Calculated weights for a given cube and coord.
+
+            Args:
+                cube : iris.cube.Cube
+                       Cube to blend across the coord.
+                coord : iris.corrd.Coord
+                        Coordinate in the cube to be blended.
+            Returns:
+                weights :  (numpy.ndarray) of weights
+                           where sum = 1.0
+                           and missing weights are set to -1.0
+                           size is the same as points in coord if
+                           points set in coord otherwise defaults to
+                           point in cube.coord
+        """
+        if not isinstance(coord, iris.coords.Coord):
+            msg = ('The second argument must be an instance of '
+                   'iris.coords.Coord but is'
+                   ' {0:s}'.format(type(coord)))
+            raise ValueError(msg)
+
+        req_coord_name = str(coord.name())
+        if not cube.coords(req_coord_name):
+            msg = ('The coord for this plugin must be '
+                   'an existing coordinate in the input cube.')
+            raise ValueError(msg)
+
+        cube_coord = cube.coord(req_coord_name)
```

Comment:
The use of `coord` and `cube_coord` is potentially a bit confusing. Are there any slightly more descriptive names that could be used?